File: Kotlin/Algorithm/src/다시 풀어본 문제/입국심사/sol2.kt
package `다시 풀어본 문제`.입국심사

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val Immigration = Array(n) { r ->
        readLine().toLong()
    }

    var left: Long = 0
    var right = (Immigration.maxOrNull()!! * m)

    while (left <right) {
        val mid = (left + right) / 2
        var total: Long = 0

        for (time in Immigration) {
            total += (mid / time)
        }

        if (total >= m) {
            right = mid
        } else {
            left = mid + 1
        }
    }
    println(left)
}

File: python/백준 온라인 저지/2022.04.20/마법사 상어와 토네이도/sol2.py
import sys
input = sys.stdin.readline

n = int(input())
sand = [list(map(int, input().split())) for _ in range(n)]
sx, sy = n//2, n//2

left = [
    (1, 1, 0.01), (-1, 1, 0.01), (1, 0, 0.07), (-1, 0, 0.07), (1, -1, 0.1),
    (-1, -1, 0.1), (2, 0, 0.02), (-2, 0, 0.02), (0, -2, 0.05), (0, -1, 0)
]

right = [(x, -y, z) for x, y, z in left]
down = [(-y, x, z) for x, y, z in left]
up = [(y, x, z) for x, y, z in left]
ans = 0

def recount_sand(time, dx, dy, direction):
    global sx, sy, ans

    for _ in range(time):
        sx += dx
        sy += dy

        if sy < 0:
            break

        total = 0

        for dx, dy, z in direction:
            nx, ny = sx + dx, sy + dy
            
            if z == 0:
                new_sand = sand[sx][sy]- total
            else:
                new_sand = int(sand[sx][sy] * z)
                total += new_sand
            
            if 0 <= nx < n and 0 <= ny < n:
                sand[nx][ny] += new_sand
            else:
                ans += new_sand

for i in range(1, n+1):
    if i % 2:
        recount_sand(i, 0, -1, left)
        recount_sand(i, 1, 0, down)
    else:
        recount_sand(i, 0, 1, right)
        recount_sand(i, -1, 0, up)

print(ans)